crates/kumod/src/smtp_server.rs

```diff
@@ -2264,6 +2280,177 @@ impl SmtpServerSession {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Diagnostic {
+                                    level: Level::ERROR,
+                                    message: "Peer Disconnected".to_string(),
+                                },
+                                when: Utc::now(),
+                            });
+                            return Err(anyhow!("peer disconnected while reading proxy protocol header"));
+                        }
+                        Ok(size) => {
+                            SmtpServerTraceManager::submit(|| SmtpServerTraceEvent {
+                                conn_meta: self.meta.clone_inner(),
+                                payload: SmtpServerTraceEventPayload::Read(data[0..size].to_vec()),
+                                when: Utc::now(),
+                            });
+                            self.read_buffer.extend_from_slice(&data[0..size]);
+                        }
+                    }
+                }
+                _ = self.shutdown.shutting_down() => {
+                    return Err(anyhow!("shutting down"));
+                }
+            };
+            let header = HeaderResult::parse(&self.read_buffer);
+            if header.is_complete() {
+                break header;
+            }
+        };
+
+        let mut addr: Option<IpAddr> = None;
+        let mut port: Option<u16> = None;
+        let mut dest_addr: Option<IpAddr> = None;
+        let mut dest_port: Option<u16> = None;
+        match header {
+            HeaderResult::V1(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v1::Addresses::Tcp4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    ppp::v1::Addresses::Tcp6(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
+                        dest_addr.replace(addresses.destination_address.into());
+                        dest_port.replace(addresses.destination_port);
+                    }
+                    _ => {}
+                }
+            }
+            HeaderResult::V2(Ok(header)) => {
+                let addresses = header.addresses;
+                match addresses {
+                    ppp::v2::Addresses::IPv4(addresses) => {
+                        addr.replace(addresses.source_address.into());
+                        port.replace(addresses.source_port);
```

Comment:
BTW, if you're setting up the integration test using our existing `ha_proxy_source_address` stuff, we set the `source_port = 0` to indicate that we don't care what source port the outgoing proxy uses for us.  I don't think that matters too much here, it will just look weird at first.